--- embedded_innodb-1.0.3.5325.orig/row/row0purge.c	2009-06-12 01:55:36.000000000 -0600
+++ embedded_innodb-1.0.3.5325/row/row0purge.c	2009-10-12 07:13:03.000000000 -0600
@@ -643,17 +643,15 @@
 
 		/* FIXME: So that purge doesn't croak on user tables. The
 		magic number 13 is the first user table. */
-		if (clust_index->cmp_ctx != NULL || node->table->id.low < 13) {
+		/* BLOB Purge patch */
+		if (node->rec_type == TRX_UNDO_DEL_MARK_REC) {
+			row_purge_del_mark(node);
+
+		} else if (updated_extern
+			   || node->rec_type ==
+			      TRX_UNDO_UPD_EXIST_REC) {
 
-			if (node->rec_type == TRX_UNDO_DEL_MARK_REC) {
-				row_purge_del_mark(node);
-
-			} else if (updated_extern
-				   || node->rec_type ==
-				      TRX_UNDO_UPD_EXIST_REC) {
-
-				row_purge_upd_exist_or_extern(node);
-			}
+			row_purge_upd_exist_or_extern(node);
 		}
 
 		if (node->found_clust) {
