diff -ur embedded_innodb-1.0.3.5325.orig/api/api0api.c embedded_innodb-1.0.3.5325/api/api0api.c
--- embedded_innodb-1.0.3.5325.orig/api/api0api.c	2009-06-12 01:55:38.000000000 -0600
+++ embedded_innodb-1.0.3.5325/api/api0api.c	2009-07-13 08:00:47.000000000 -0600
@@ -4051,6 +4051,7 @@
 	case DATA_BLOB:
 	case DATA_VARCHAR:
 	case DATA_VARCLIENT:
+	case DATA_BINARY:
 		memcpy(dst, src, len);
 		break;
 
