Index: row/row0sel.c
===================================================================
--- embedded_innodb-1.0.3.5325.orig/row/row0sel.c	(revision 6143)
+++ embedded_innodb-1.0.3.5325/row/row0sel.c	(working copy)
@@ -3790,7 +3790,11 @@ requires_clust_rec:
 		}
 
 		goto next_rec;
-	}
+
+	} 
+
+	ut_a(!row_sel_row_cache_is_full(prebuilt));
+	row_sel_row_cache_add(prebuilt, result_rec, offsets);
 
 	/* From this point on, 'offsets' are invalid. */
 got_row:
